scalyr_agent/monitor_utils/k8s.py

```diff
@@ -566,42 +580,209 @@ def process_object( self, obj ):
+
+        self._lock = threading.Lock()
+        self._namespace_filter = None
+        self._node_filter = None
+        self._k8s = None
+        self._cache_expiry_secs=cache_expiry_secs
+        self._cache_purge_secs=cache_purge_secs
+        self._config_update_count = 0
+
+        empty_config = _CacheConfig( api_url='', namespaces_to_ignore=[] )
 
-        self._k8s = k8s
 
-        self._namespace_filter = self._build_namespace_filter( namespaces_to_ignore )
-        self._node_filter = self._build_node_filter( self._namespace_filter )
+        new_config = _CacheConfig( api_url=api_url,
+                                   verify_api_queries=verify_api_queries,
+                                   cache_expiry_secs=cache_expiry_secs,
+                                   cache_purge_secs=cache_purge_secs,
+                                   namespaces_to_ignore=namespaces_to_ignore )
+
+        # perform initial configuration of the cache
+        self._configure( old_config=empty_config, new_config=new_config )
 
         # create the controller cache
-        self._controller_processor = ControllerProcessor( logger )
-        self._controllers = _K8sCache( logger, k8s, self._controller_processor, '<controller>',
-                               filter=self._namespace_filter,
+        self._controller_processor = ControllerProcessor()
+        self._controllers = _K8sCache( self._controller_processor, '<controller>',
                                perform_full_updates=False )
 
         # create the pod cache
-        self._pod_processor = PodProcessor( logger, self._controllers )
-        self._pods = _K8sCache( logger, k8s, self._pod_processor, 'Pod',
-                               filter=self._node_filter )
+        self._pod_processor = PodProcessor( self._controllers )
+        self._pods = _K8sCache( self._pod_processor, 'Pod' )
 
         self._cluster_name = None
-        self._cache_expiry_secs = cache_expiry_secs
-        self._cache_purge_secs = cache_purge_secs
         self._last_full_update = time.time() - cache_expiry_secs - 1
 
-        self._lock = threading.Lock()
         self._initialized = False
 
-        self._thread = StoppableThread( target=self.update_cache, name="K8S Cache" )
+        self._thread = None
+
+        if start_caching:
+            self.start()
+
+    def _configure( self, old_config, new_config ):
+        """
+        Configures the cache based on any changes in the configuration
+        @param old_config: the old configuration
+        @param new_config: the new configuration
+        @type old_config: _CacheConfig
+        @type new_config: _CacheConfig
+        @return: True if the config changed, False otherwise
+        """
+
+        changed = False
+
+        # get old state values
+        old_k8s = None
+        old_namespace_filter = None
+        old_node_filter = None
+        old_config_update_count = 0
+        self._lock.acquire()
+        try:
+            old_k8s = self._k8s
```

Comment:
I believe there are only two ways we will not return True in this method.  Either, the two configurations are the same so no change is needed... or this thread does not end up having its configuration recognized because some other thread updated the configuration over it.

I think you can simplify your code by checking right here (while you hold the lock) if the new configuration is equal to `self._config`.  If it is, then you can just return False here.  Then.. the only way you could return False is if your thread doesn't win (which we will handle down below).

So, to ease this, you should have an equals method on the CacheConfig object.

Reply:
Done.  Added both __eq__ and __ne__ (because __ne__ is not defined by default based on __eq__).